db/services.php

```diff
@@ -128,6 +128,27 @@
         'description' => 'Return the final course grade for the given users',
         'type'        => 'read'
     ),
+    'local_aspiredu_get_users_by_roles' => [
+        'classname'    => 'local_aspiredu\external\get_users_by_roles',
+        'classpath'    => '',
+        'description'  => 'Return a list of users given a list of roles',
+        'type'         => 'read',
+        'capabilities' => 'moodle/user:viewdetails, moodle/user:viewhiddendetails, moodle/course:useremail'
+    ],
+    'local_aspiredu_get_users_by_capabilities' => [
+        'classname'    => 'local_aspiredu\external\get_users_by_capabilities',
+        'classpath'    => '',
+        'description'  => 'Return a list of users given a list of capabilities',
+        'type'         => 'read',
+        'capabilities' => 'moodle/user:viewdetails, moodle/user:viewhiddendetails, moodle/course:useremail'
+    ],
+    'local_aspiredu_get_courses' => [
+            'classname' => 'local_aspiredu\external\get_courses',
+            'classpath' => '',
+            'description' => 'Return paginated courses details',
+            'type' => 'read',
+            'capabilities' => 'moodle/course:view, moodle/course:update, moodle/course:viewhiddencourses',
```

Comment:
@guillogo when attempting to install this version of the plugin on our test moodle instance, I get the error below. I believe it's because these don't define the `methodname` property.

```
Debug info: Column 'methodname' cannot be null
INSERT INTO mdl_external_functions (name,classname,methodname,classpath,component,capabilities,services) VALUES(?,?,?,?,?,?,?)
-- line 1403 of /lib/dml/mysqli_native_moodle_database.php: call to mysqli_native_moodle_database->insert_record_raw()
-- line 1276 of /lib/upgradelib.php: call to mysqli_native_moodle_database->insert_record()
-- line 711 of /lib/upgradelib.php: call to external_update_descriptions()
-- line 1922 of /lib/upgradelib.php: call to upgrade_plugins()
-- line 713 of /admin/index.php: call to upgrade_noncore()
[array (
0 => 'local_aspiredu_get_users_by_roles',
1 => 'local_aspiredu\\external\\get_users_by_roles',
2 => NULL,
3 => NULL,
4 => 'local_aspiredu',
5 => 'moodle/user:viewdetails, moodle/user:viewhiddendetails, moodle/course:useremail',
6 => NULL,
)]
Error code: dmlwriteexception 
```

```
Stack trace:

    line 489 of /lib/dml/moodle_database.php: dml_write_exception thrown
    line 1357 of /lib/dml/mysqli_native_moodle_database.php: call to moodle_database->query_end()
    line 1403 of /lib/dml/mysqli_native_moodle_database.php: call to mysqli_native_moodle_database->insert_record_raw()
    line 1276 of /lib/upgradelib.php: call to mysqli_native_moodle_database->insert_record()
    line 711 of /lib/upgradelib.php: call to external_update_descriptions()
    line 1922 of /lib/upgradelib.php: call to upgrade_plugins()
    line 713 of /admin/index.php: call to upgrade_noncore()
```

Reply:
Setting it to an empty string seems to have allowed it to save. If you agree with that solution, please resolve this conversation.

Reply:
hi @tim-schilling I'd use as `'methodname'  => 'execute',` instead. Also, what Moodle version do you have in your test moodle instance?

cheers

Reply:
No worries the last commit with the 'execute' will do the trick for 3.10